core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/FnSubstring.java

```diff
@@ -0,0 +1,180 @@
+      .deterministic()
+      .contextIndependent()
+      .focusIndependent()
+      .argument(IArgument.builder()
+          .name("sourceString")
+          .type(IStringItem.class)
+          .zeroOrOne()
+          .build())
+      .argument(IArgument.builder()
+          .name("start")
+          .type(IDecimalItem.class)
+          .one()
+          .build())
+      .argument(IArgument.builder()
+          .name("length")
+          .type(IDecimalItem.class)
+          .one()
+          .build())
+      .returnType(IStringItem.class)
+      .returnOne()
+      .functionHandler(FnSubstring::executeThreeArg)
+      .build();
+
+  private FnSubstring() {
+    // disable construction
+  }
+
+  @SuppressWarnings("unused")
+  @NonNull
+  private static ISequence<IStringItem> executeTwoArg(
+      @NonNull IFunction function,
+      @NonNull List<ISequence<?>> arguments,
+      @NonNull DynamicContext dynamicContext,
+      IItem focus) {
+
+    IStringItem sourceString = FunctionUtils.asTypeOrNull(arguments.get(0).getFirstItem(true));
+
+    if (sourceString == null) {
+      return ISequence.of(IStringItem.valueOf(""));
+    }
+
+    IDecimalItem start = FunctionUtils.asType(ObjectUtils.requireNonNull(arguments.get(1).getFirstItem(true)));
+
+    int startIndex = start.round().asInteger().intValue();
+    int endIndex = (sourceString.toString().length() - startIndex) + 1;
+
+    return ISequence.of(fnSubstring(sourceString, startIndex, endIndex));
+  }
+
+  @SuppressWarnings("unused")
+  @NonNull
+  private static ISequence<IStringItem> executeThreeArg(
+      @NonNull IFunction function,
+      @NonNull List<ISequence<?>> arguments,
+      @NonNull DynamicContext dynamicContext,
+      IItem focus) {
+
+    IStringItem sourceString = FunctionUtils.asTypeOrNull(arguments.get(0).getFirstItem(true));
+
+    if (sourceString == null) {
+      return ISequence.of(IStringItem.valueOf(""));
+    }
+
+    IDecimalItem start = FunctionUtils.asType(ObjectUtils.requireNonNull(arguments.get(1).getFirstItem(true)));
+    IDecimalItem length = FunctionUtils.asType(ObjectUtils.requireNonNull(arguments.get(2).getFirstItem(true)));
+
+    int startIndex = start.round().asInteger().intValue();
+    int endIndex = length.round().asInteger().intValue();
+
+    return ISequence.of(fnSubstring(sourceString, startIndex, endIndex));
+  }
+
+  /**
+   * An implementation of XPath 3.1 <a href=
+   * "https://www.w3.org/TR/xpath-functions-31/#func-substring">fn:substring</a>.
+   *
+   * @param sourceString
+   *
+   * @param start
+   *
+   * @param length
+   *
+   * @return the atomized result
+   */
+  @NonNull
+  public static IStringItem fnSubstring(@NonNull IStringItem sourceString, @NonNull IDecimalItem start,
+      @NonNull IDecimalItem length) {
+
+    int startIndex = start.round().asInteger().intValue();
+    int endIndex = length.round().asInteger().intValue();
+    return fnSubstring(sourceString, startIndex, endIndex);
+  }
```

Comment:
Is this version used?

Reply:
We got rid of this during our last pairing session.